src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -70,23 +70,65 @@ def _learn(self) -> List[PartialNSRTAndDatastore]:
-            for pnads in param_opt_to_nec_pnads.values():
-                for pnad in pnads:
-                    pnad.poss_keep_effects.clear()
-                    pnad.seg_to_keep_effects_sub.clear()
-            # Run one pass of backchaining.
-            nec_pnad_set_changed = self._backchain_one_pass(
-                param_opt_to_nec_pnads, param_opt_to_general_pnad)
-
-        final_pnads = self._finish_learning(param_opt_to_nec_pnads)
+        curr_finalized_pnads: Set[PartialNSRTAndDatastore] = set()
+        finalized_pnads_converged = False
+
+        while not finalized_pnads_converged:
+            # Pass over the demonstrations multiple times. Each time, backchain
+            # to learn PNADs. Repeat until a fixed point is reached.
+            nec_pnad_set_changed = True
+            while nec_pnad_set_changed:
+                # Before each pass, clear the poss_keep_effects
+                # of all the PNADs. We do this because we only want the
+                # poss_keep_effects of the final pass, where the PNADs did
+                # not change. However, we cannot simply clear the the
+                # sub_dict because some of these substitutions might be
+                # necessary if this happens to be a PNAD that already
+                # has keep effects. Thus, we call a method that handles
+                # this correctly.
+                for pnads in param_opt_to_nec_pnads.values():
+                    for pnad in pnads:
+                        pnad.poss_keep_effects.clear()
+                        self._clear_unnecessary_keep_effs_sub(pnad)
+                # Run one pass of backchaining.
+                nec_pnad_set_changed = self._backchain_one_pass(
+                    param_opt_to_nec_pnads, param_opt_to_general_pnad)
+
+            # Induce delete effects, side predicates and potentially
+            # keep effects.
+            self._finish_learning(param_opt_to_nec_pnads)
+
+            # Recompute datastores and preconditions for all PNADs.
+            new_finalized_pnads = [
+                pnad for pnad_list in param_opt_to_nec_pnads.values()
+                for pnad in pnad_list
+            ]
+            self._recompute_datastores_from_segments(new_finalized_pnads)
+            new_pnads_with_datastores = []
+            for pnad in new_finalized_pnads:
+                if len(pnad.datastore) > 0:
+                    pnad.op = pnad.op.copy_with(
+                        preconditions=self.
+                        _induce_preconditions_via_intersection(pnad))
+                    new_pnads_with_datastores.append(pnad)
+                else:
+                    param_opt_to_nec_pnads[pnad.option_spec[0]].remove(pnad)
+
+            new_finalized_pnads = new_pnads_with_datastores
+
+            # Check if the finalized PNAD set has converged.
+            if set(new_finalized_pnads) == curr_finalized_pnads:
+                finalized_pnads_converged = True
```

Comment:
i don't think you need this `finalized_pnads_converged` variable... can't you just do a `break` here instead?

Reply:
fixed